### DISCUSSION TITLE
Force google engine to only use ipv4 (Hetzner IPv6 & google)

### DISCUSSION THREAD

**Discussion by charonfaustinus** [searxng/searxng#3188] [Q&A] [answered]
I have been using searxng for a long time, but for some reason google blocked my server from accessing ipv6 of the site.
My test is try going into google.com on ipv6 using w3m and I get this page.

![Termius_9CALnHnBc5](https://github.com/searxng/searxng/assets/136845818/ab585bfd-2a39-4d8d-a051-b8acb468e743)

But then, when I try going to google.com with forcing ipv4 (`w3m google.com -4`) I get access to google normally.
![Termius_uM7AC9yCn2](https://github.com/searxng/searxng/assets/136845818/436c0fa0-00a7-41fa-acb7-3cd50a0f9cb4)

Is there a way to force searxng to only use ipv4 with google engine without losing the ability to serve ipv6 to my instance of searxng?

**Comment by return42** [comment, marked as answer]
The bot defense of the original search engines tends to react more sensitively to IPv6 / I recommend to route all outgoing requests over IPv4:

```yaml
outgoing:
  source_ips: 0.0.0.0
```

--> https://docs.searxng.org/admin/settings/settings_outgoing.html

-----

Update: I have a suspicion that google tend to block some of Hetzner's IPv6 networks:

- https://docs.saltbox.dev/faq/Hetzner/
- https://support.google.com/websearch/thread/227891804/http-403-errors-when-using-google-search-from-hetzner-ipv6-addresses?hl=en

**Reply by unixfox** [reply]
bot protections works the same way as ipv4 and ipv6. it's just that in ipv6 the bot protection probably block an entire ipv6 range where as in ipv4 probably a couple of ip addresses.

**Comment by unixfox** [comment]
I would suggest rotating your IPv6 address first. Sometimes it's not the entire ipv6 range that is blocked.

**Reply by return42** [reply]
@unixfox Do you see advantages in using IPv6 for outgoing requests? ... I don't see any advantage / not sure .. maybe I'm missing something?

**Reply by unixfox** [reply]
you should ask the question instead this way:
```
why do I need to use ipv4 when ipv6 is working fine?
```

ipv4 exists due to legacy reasons. ipv6 is the future, it exists because there are not enough public ipv4 addresses in the world for everyone.

it's time to stop the old way of thinking: I prefer to use ipv4. It's because a lot of people think this way that ipv6 is slow to deploy around the world.
people have to understand that (most of) their stuff will work in ipv6, and they need to make a little more effort in not trying to exclusively use ipv4.

**Reply by return42** [reply]
OK, thanks for feedback / I make it a little easier for myself and look at it from the practical side ... and there I have to realize that IPv4 is better suited for obfuscation before bot detection (at least in the use-case we have here)

**Reply by charonfaustinus** [reply]
![Termius_nYry3LFmbA](https://github.com/searxng/searxng/assets/136845818/de46ed9d-7cf0-4972-bc85-4dcf453f12e8)
I tried using [smart-ipv6-rotator](https://github.com/iv-org/smart-ipv6-rotator), and I get a new ipv6 address but still got this page and too many request from google engine. However, enforcing searxng to use ipv4 work fine. Already retired for 3 times with 3 different IPv6. The result are the same.

**Reply by unixfox** [reply]
then it's the entire ipv6 range that is blocked. once google detect that searxng is also running on your ipv4 address then it will get blocked too.
it doesn't matter if it's ipv4 or ipv6

**Reply by return42** [reply]
@mahehro @charonfaustinus if you get blocked again by google and outgoing requests on IPv4 please report it here / thanks :+1: 

@charonfaustinus is your SearXNG instance running on Hetzner?

**Reply by charonfaustinus** [reply]
Yes, It was hosted on Hetzner, at US east location

**Reply by return42** [reply]
I have a suspicion that google tend to block some of Hetzner's IPv6 networks:

- https://docs.saltbox.dev/faq/Hetzner/
- https://support.google.com/websearch/thread/227891804/http-403-errors-when-using-google-search-from-hetzner-ipv6-addresses?hl=en
